### PR TITLE
PIM-11062 Fix category filter behavior in product grid

### PR DESCRIPTION
**Pull request by amigaire**
**Description (for Contributor and Core Developer)**

The endpoint to fetch all category trees, their children and the selected category filter only used the id of the selected category.
But because the filter use the value -1 for "Unclassified products" and -2 for "All products", the endpoint could not retrieve the category tree linked to a children category and it considered the configured user's default tree as the current selected category.

The fix add the selected category tree id to the data send to the endpoint, so it can return the correct selected filter value to the front.

Corrected behavior :

[Screencast from 27-06-2023 10:49:03.webm](https://github.com/akeneo/pim-community-dev/assets/94356747/db9271fa-6aa2-4a40-b9d4-48f10f093e88)

**Definition Of Done (for Core Developer only)**

- [x] Tests
- [x] Changelog (maintenance bug fixes)

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Component/Category/CategoryTree/UseCase/ListRootCategoriesWithCountHandler.php [hunk 1]
```diff
@@ -53,29 +54,67 @@ public function __construct(
...
+        $defaultCategoryTreeId = $this->userContext->getAccessibleUserTree()?->getId();
+        if (null !== $query->categoryTreeIdSelectedAsFilter()) {
+            $defaultCategoryTreeId = $query->categoryTreeIdSelectedAsFilter();
+        } else if (null !== $query->categoryIdSelectedAsFilter()) {
+            $selectedCategory = $this->categoryRepository->find($query->categoryIdSelectedAsFilter());
+            if (null !== $selectedCategory && null !== $selectedCategory->getRoot()) {
+                $defaultCategoryTreeId = $selectedCategory->getRoot();
+            }
+        }
+        return $defaultCategoryTreeId;
```
Comment (pchasle) [inline src/Akeneo/Pim/Enrichment/Component/Category/CategoryTree/UseCase/ListRootCategoriesWithCountHandler.php:102]: Suggestion to use early return, easier to read & match the comment steps order.

```suggestion
    if (null !== $query->categoryTreeIdSelectedAsFilter()) {
        return $query->categoryTreeIdSelectedAsFilter();
    }

    if (null !== $query->categoryIdSelectedAsFilter()) {
        $selectedCategory = $this->categoryRepository->find($query->categoryIdSelectedAsFilter());
        if (null !== $selectedCategory && null !== $selectedCategory->getRoot()) {
            return $selectedCategory->getRoot();
        }
    }

    return $this->userContext->getAccessibleUserTree()?->getId();
```

File: src/Akeneo/Pim/Enrichment/Component/Category/CategoryTree/UseCase/ListRootCategoriesWithCountHandler.php [hunk 2]
```diff
@@ -53,29 +54,67 @@ public function __construct(
...
+            if (null !== $defaultCategoryTreeId) {
+                // selected category does not exist in DB, so we get the categoryTree from the default categoryTreeId
+                $selectedCategory = $this->categoryRepository->find($defaultCategoryTreeId);
+            }
+            if ($selectedCategory === null) {
+                // we can't find a proper category tree
+                return null;
+            }
+        }
+        return $this->categoryRepository->find($selectedCategory->getRoot());
```
Comment (pchasle) [inline src/Akeneo/Pim/Enrichment/Component/Category/CategoryTree/UseCase/ListRootCategoriesWithCountHandler.php:118]: Same suggestion.

```suggestion
        $selectedCategory = $this->categoryRepository->find($subCategoryId);
        if (null !== $selectedCategory) {
            return $this->categoryRepository->find($selectedCategory->getRoot());
        }

        // selected category does not exist in DB, so we get the categoryTree from the default categoryTreeId
        if (null !== $defaultCategoryTreeId) {
            return $this->categoryRepository->find($defaultCategoryTreeId);
        }

        return null;
```

**Review by pchasle — APPROVED** [review]

File: src/Akeneo/Pim/Enrichment/Component/Category/CategoryTree/UseCase/ListRootCategoriesWithCountHandler.php [hunk 3]
```diff
@@ -53,29 +54,68 @@ public function __construct(
...
-                return [];
-            }
+        $defaultCategoryTreeId = $this->getDefaultCategoryTreeId($query);
+        if ($defaultCategoryTreeId === null) {
+            return [];
         }
-        $rootCategoryIdToExpand = $categorySelectedAsFilter->getRoot();
 
-        $rootCategories = $query->countIncludingSubCategories() ?
+        $selectedCategoryTree = $this->getCategoryTreeFromSelectedSubCategory(
```
Comment (amigaire) [inline src/Akeneo/Pim/Enrichment/Component/Category/CategoryTree/UseCase/ListRootCategoriesWithCountHandler.php:62]:
```suggestion
        $rootCategoryToExpand = $this->getRootCategoryToExpandFromSelectedCategory(
```

File: src/Akeneo/Pim/Enrichment/Component/Category/CategoryTree/UseCase/ListRootCategoriesWithCountHandler.php [hunk 4]
```diff
@@ -53,29 +54,68 @@ public function __construct(
...
+    /**
+     * Try to get the categoryTreeId from url parameters.
+     * If not set, try to get the root category id from selectedCategoryId in url parameters.
+     * If the root category does not exist, try to get the user accessible categoryTreeId.
+     * @param ListRootCategoriesWithCount $query
+     * @return int|null
+     */
+    private function getDefaultCategoryTreeId(ListRootCategoriesWithCount $query): ?int
+    {
+    if (null !== $query->categoryTreeIdSelectedAsFilter()) {
```
Comment (amigaire) [inline src/Akeneo/Pim/Enrichment/Component/Category/CategoryTree/UseCase/ListRootCategoriesWithCountHandler.php:93]: linter !

**Review by etiennekerbastard — APPROVED** [review]

**Review by jeremyBeaucousinAkeneo — APPROVED** [review]

**Review by mmetayer — APPROVED** [review]
thanks @amigaire !

**Pull request merged**